ts/client/src/client.ts

```diff
@@ -3374,64 +3372,230 @@ export class MangoClient {
+      thresholdPriceInSellPerBuyToken,
+      maxBuyUi ?? -account.getTokenBalanceUi(buyBank),
+      maxSellUi ?? Number.MAX_SAFE_INTEGER,
+      'TakeProfitOnBorrow',
+      pricePremium,
+      allowCreatingDeposits ?? false,
+      allowCreatingBorrows ?? false,
+      expiryTimestamp,
+    );
+  }
+
+  public async tcsStopLossOnBorrow(
+    group: Group,
+    account: MangoAccount,
+    sellBank: Bank,
+    buyBank: Bank,
+    thresholdPriceUi: number,
+    thresholdPriceInSellPerBuyToken: boolean,
+    maxBuyUi: number | null,
+    maxSellUi: number | null,
+    pricePremium: number | null,
+    allowCreatingDeposits: boolean | null,
+    allowCreatingBorrows: boolean | null,
+    expiryTimestamp: number | null,
+  ): Promise<TransactionSignature> {
+    if (maxBuyUi == null && account.getTokenBalanceUi(sellBank) > 0) {
+      throw new Error(
+        `Only allowed to set stop loss on borrows! Current balance ${account.getTokenBalanceUi(
+          buyBank,
+        )}`,
+      );
+    }
+
+    return await this.tokenConditionalSwapCreate(
+      group,
+      account,
+      sellBank,
+      buyBank,
+      thresholdPriceUi,
+      thresholdPriceInSellPerBuyToken,
+      maxBuyUi ?? -account.getTokenBalanceUi(buyBank),
+      maxSellUi ?? Number.MAX_SAFE_INTEGER,
+      'StopLossOnBorrow',
+      pricePremium,
+      allowCreatingDeposits ?? false,
+      allowCreatingBorrows ?? false,
+      expiryTimestamp,
+    );
+  }
+
+  public async tokenConditionalSwapCreate(
+    group: Group,
+    account: MangoAccount,
+    sellBank: Bank,
+    buyBank: Bank,
+    thresholdPriceUi: number,
+    thresholdPriceInSellPerBuyToken: boolean,
+    maxBuyUi: number,
+    maxSellUi: number,
+    tcsIntention:
+      | 'TakeProfitOnDeposit'
+      | 'StopLossOnDeposit'
+      | 'TakeProfitOnBorrow'
+      | 'StopLossOnBorrow'
+      | null,
+    pricePremium: number | null,
+    allowCreatingDeposits: boolean,
+    allowCreatingBorrows: boolean,
+    expiryTimestamp: number | null,
+  ): Promise<TransactionSignature> {
+    const maxBuy = toNative(maxBuyUi, buyBank.mintDecimals);
+    const maxSell = toNative(maxSellUi, sellBank.mintDecimals);
+
+    if (!thresholdPriceInSellPerBuyToken) {
+      thresholdPriceUi = 1 / thresholdPriceUi;
+    }
+
+    let lowerLimit, upperLimit;
+    const thresholdPrice = toNativeSellPerBuyTokenPrice(
+      thresholdPriceUi,
+      sellBank,
+      buyBank,
     );
-    pricePremiumFraction =
-      pricePremiumFraction > 0 ? pricePremiumFraction : 0.3;
+    const sellTokenPerBuyTokenPrice = buyBank.price
+      .div(sellBank.price)
+      .toNumber();
+
+    if (
+      tcsIntention == 'TakeProfitOnDeposit' ||
+      tcsIntention == 'StopLossOnBorrow' ||
+      (tcsIntention == null && thresholdPrice > sellTokenPerBuyTokenPrice)
+    ) {
+      lowerLimit = thresholdPrice;
+      upperLimit = Number.MAX_SAFE_INTEGER;
+    } else {
+      lowerLimit = 0;
+      upperLimit = thresholdPrice;
+    }
```

Comment:
minor: I was surprised to see this here instead of in the wrapper functions? They know what they want and could pass us the appropriate upper/lower limit?

Reply:
I think like 75% of below code will be repeated, hence I did it here

```
if (!thresholdPriceInSellPerBuyToken) {
      thresholdPriceUi = 1 / thresholdPriceUi;
    }

    let lowerLimit, upperLimit;
    const thresholdPrice = toNativeSellPerBuyTokenPrice(
      thresholdPriceUi,
      sellBank,
      buyBank,
    );
    const sellTokenPerBuyTokenPrice = buyBank.price
      .div(sellBank.price)
      .toNumber();

    if (
      tcsIntention == 'TakeProfitOnDeposit' ||
      tcsIntention == 'StopLossOnBorrow' ||
      (tcsIntention == null && thresholdPrice > sellTokenPerBuyTokenPrice)
    ) {
      lowerLimit = thresholdPrice;
      upperLimit = Number.MAX_SAFE_INTEGER;
    } else {
      lowerLimit = 0;
      upperLimit = thresholdPrice;
    }
```